async-book/src/async_await/chapter.md

```diff
@@ -0,0 +1,154 @@
+up by the executor and will resume running, allowing the `await!` to resolve.
+
+## `async` Lifetimes
+
+Unlike traditional functions, `async fn`s which take references or other
+non-`'static` arguments return a `Future` which is bounded by the lifetime of
+the arguments:
+
+```rust
+// This function:
+async fn foo(x: &u8) -> u8 { *x }
+
+// Is equivalent ot this function:
+fn foo<'a>(x: &'a u8) -> impl Future<Output = ()> + 'a {
+    async { *x }
+}
+```
+
+This means that the future returned from an `async fn` must be `await!`ed
+while its non-`'static` arguments are still valid. In the common
+case of `await!`ing the future immediately after calling the function
+(like `await!(foo(&x))`) this is not an issue. However, if storing the future
+or sending it over to another task or thread, this may be an issue.
+
+One common workaround for turning an `async fn` with references-as-arguments
+into a `'static` future is to bundle the arguments with the call to the
+`async fn` inside an `async` block:
+
+```rust
+async fn foo(x: &u8) -> u8 { *x }
+
+fn bad() -> impl Future<Output = ()> {
+    let x = 5;
+    foo(&x) // ERROR: `x` does not live long enough
+}
+
+fn good() -> impl Future<Output = ()> {
+    async {
+        let x = 5;
+        await!(foo(&x))
+    }
+}
+```
+
+By moving the argument into the `async` block, we extend its lifetime to match
+that of the `Future` returned from the call to `foo`.
+
+## `async move`
+
+`async` blocks and closures allow the `move` keyword, much like normal
+closures. An `async move` body will take ownership of the variables it
+references, allowing it to outlive the current scope, but giving up the ability
+to share those variables with other code:
+
+```rust
+/// `async` block:
+///
+/// Multiple different `async` blocks can access the same local variable
+/// so long as they're executed within the variable's scope.
+async fn foo() {
+    let my_string = "foo".to_string();
+
+    let future_one = async {
+        ...
+        println!("{}", my_string);
+    };
+
+    let future_two = async {
+        ...
+        println!("{}", my_string);
+    };
+ 
+    // Run both futures to completion, printing "foo" twice
+    let ((), ()) = join!(future_one, future_two);
+}
+
+/// `async move` block:
+///
+/// Only one `async` block can access captured variables, since they are
+/// moved into the `Future` generated by the `async` block. However,
+/// this allows the `Future` to outlive the original scope of the variable:
+fn foo() -> impl Future<Output = ()> {
+    let my_string = "foo".to_string();
+    async move {
+        ...
+        println!("{}", my_string);
+    }
+}
+```
+
+## `await!`ing on a Multithreaded Executor
+
+Note that, when using a multithreaded `Future` executor, a `Future` may move
+between threads, so any variables used in `async` bodies must be able to travel
+between threads, as any `await!` can potentially result in a switch to a new
+thread.
```

Comment:
This sentence spans 3 lines. Breaking it up into a few smaller sentences might help make it a bit easier to read.